docs/generatedcodeguide.rst

```diff
@@ -0,0 +1,228 @@
+=====================
+Chapel Generated Code
+=====================
+
+This page describes exactly what Chapel code the protocol buffer compiler
+generates for protocol definitions using ``proto3`` syntax. You should read
+the `proto3 language guide`_ before reading this document.
+
+Compiler Invocation
+-------------------
+The protocol buffer compiler produces Chapel output when invoked with the ``--chpl_out``
+command-line flag. The parameter to the ``--chpl_out`` option is the directory where
+you want the compiler to write your Chapel output. The compiler creates a single
+source file for each ``.proto`` file input, having a ``.chpl`` extension.
+
+Only ``proto3`` messages are supported by the Chapel code generator. Ensure
+that each ``.proto`` file begins with a declaration of:
+
+.. code-block:: proto
+
+  syntax = "proto3";
+
+Ouput module name
+-----------------
+The name of the ouput file/module will be same as the ``package`` name. If the
+``package`` name is not specified, the module takes the name of the proto
+file with all non-aplhanumeric characters replaced by an ``underscore``.
+
+For example a file called ``address.proto`` without the package specifier will
+result in an ouput file called ``address.chpl`` with the generated code wrapped
+in a module of the same name.(Full implementation is not shown here.)
+
+.. code-block:: chpl
+
+  module address {
+    record messageName {
+    ...
+    }
+  }
+
+The same proto file with the a package declaration ``package myPackage;`` will
+result in a file called ``myPackage.chpl``.
+
+.. code-block:: chpl
+
+  module myPackage {
+    record messageName {
+    ...
+    }
+  }
+
+Messages
+--------
+Given a simple message declaration:
+
+.. code-block:: proto
+  
+  message Foo {
+    int32 num = 1;
+  }
+  
+The protocol buffer compiler generates a record called ``Foo``, which have message
+field initializers and serialization/parsing methods for wire-type encoding.
+
+ .. code-block:: chpl
+  
+  record Foo {
+
+    // Record fields will be generated corresponding to each proto message field.
+    var num: int(32);
+    
+    // Used to store encoded byte stream of unknown fields encountered while parsing.
+    // As per proto3 documentation, unknown fields should be preserved and appended
+    // to the generated message byte stream.
+    var unknownFieldStream: bytes = "";
+    
+    // User exposed method for serializing data to protobuf wire format.
+    // This is a wrapper method to the actual method, it is used for making a
+    // binary writing channel out of the channel given as an input by the user. 
```

Comment:
I guess we can keep it since this guide is not only for users. Some chapel developers might be interested in knowing the details of the generated code and this guide might be helpful in that case. 

(This guide might have helped in the API review process as well. A link to this guide might have saved us from some comment explanation :P)

What do you say?

Reply:
The trouble is that describing details like this can make it harder to change the implementation in the future - anything you talk about explicitly can be relied upon and thus broken if you reshuffle things.  This seems to me like a detail that is relevant when reading the code itself, and also something it would be easy to change there without realizing that this document needed to be updated as well.

Reply:
Hmm...Makes sense, I guess this should be enough then? 

`User exposed method for serializing data to protobuf wire format. This is a wrapper method to the actual method.`